### PR TITLE
Comp player

### PR DESCRIPTION
**Pull request by lpenzey**

### CODE REVIEW

File: lib/TTT/comp_player.ex
Comment (saralein) [inline lib/TTT/comp_player.ex]: I would take a look at these regarding module names:

- https://github.com/elixir-lang/elixir/blob/master/lib/elixir/pages/Naming%20Conventions.md

- https://github.com/christopheradams/elixir_style_guide#modules

You might also consider a less abbreviated name.

Comment (lpenzey) [reply lib/TTT/comp_player.ex]: 👍 thanks!

File: lib/TTT/input.ex
Comment (esmevane) [inline lib/TTT/input.ex:28]: This is interesting.  Will this check each valid case, and then only progress to placing moves if they all pass, otherwise it's just a no-op?  Is that a correct read?

File: lib/TTT/validity.ex
Comment (esmevane) [inline lib/TTT/validity.ex]: This coercion from string to integer could possibly happen in your `input` module, or maybe before the move gets passed to validity.  Validity could potentially be better off in a world where it can trust the types of the things being handed to it.

File: lib/TTT/validity.ex
Comment (esmevane) [inline lib/TTT/validity.ex]: Same as above, this might be something you're only needing to do in here because you have an opportunity to sort of pre-shape `move` into an integer before `validity` gets it.

For example, what happens if move has to become more complex?  That may not happen in this project, so for the sake of a thought exercise, at least.  If `move` has to become like, a struct or something, or carry some additional information around (maybe we want to start tracking _when_ a move happens, as a random thought) - what changes need to happen to acommodate that, and where?

If we have a scenario where `move` is turned into the type of value we need it to be, most ideally at the time _when_ we get it, then that's straightforward: we just augment `move` itself, and maybe change how we're marshaling it.

If we have a scenario where `move` stays as a primitive value (in this case, a String) and everything reshapes itself so that it can work with Strings, then that sort of change is a lot more intensive in terms of the scope of work we need to do.  So, our code in that case would sort of resist that change.

File: test/comp_player_test.exs
Comment (esmevane) [inline test/comp_player_test.exs]: What is the elixir casing convention, here?  Is it `StartCase` or `Snake_Case`?  It won't necessarily harm your code to split off from a language's conventions, but it's usually encouraged to find the language-level idiom for casing and naming, and stick to that (especially for public interfaces)

File: test/input_test.exs
Comment (esmevane) [inline test/input_test.exs]: It looks like our elixir idiom is `StartCase`

File: test/state_test.exs
Comment (esmevane) [inline test/state_test.exs]: There are a few exceptions to this, but those aside, we will almost always prefer descriptive variable names over single-character ones.  Since this is a test, I'd even say go a bit further than that, with the variable names.  This should be as easy to read as possible for a future developer, so like, maybe something like:

`empty_board`

...Or something like that.

Then line 82 would look like `assert State.set_move(move, empty_board)`.  If I see that in test output, or I hop directly to that line, then it gives me a good impression of what's going on _without looking around the rest of the file first_.  That's a huge boost in terms of cognitive help for new readers, or even yourself after you've slept a few nights and worked on other things.

**Review by esmevane — CHANGES_REQUESTED** [review]
I think this was overall good, but I'd like to see some changes before I sign off entirely.  I think you should take a look at your module names and variable names (it seemed like the variable names in tests, mainly).

Another thing I've brought up is the idea of rethinking your `move` value and how the rest of your program interacts with it.  The idea here is that we can sort of shape our code to be more flexible and maintainable if we guard our boundaries a little.

In this case, "guarding our boundaries" means something along the lines of taking the `String.to_integer` call and just doing that when `move` enters the program, vs. other spots where it might be used.

[This is a great talk on boundaries and how to recognize and leverage them](https://www.youtube.com/watch?v=yTkzNHF6rMs).

File: lib/TTT/input.ex
Comment (lpenzey) [reply lib/TTT/input.ex:28]: Yes that's correct - if the patterns on each side of the arrow are a match it will proceed to the do block, if one of the lines doesn't match it instead returns the non-matching pattern, which in this case is the specific error message such as {:error, :not_a_number}

**Pull request merged**
